tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,753 @@
+	}
+
+	// Encoding.
+	chkEnc := sgmBytes.Range(chkStart, chkStart+ChunkEncodingSize)[0]
+
+	// Data length.
+	// With the minimum chunk length this should never cause us reading
+	// over the end of the slice.
+	chkDataLenStart := chkStart + ChunkEncodingSize
+	c := sgmBytes.Range(chkDataLenStart, chkDataLenStart+MaxChunkLengthFieldSize)
+	chkDataLen, n := binary.Uvarint(c)
+	if n <= 0 {
+		return nil, errors.Errorf("reading chunk length failed with %d", n)
+	}
+
+	// Verify the chunk data end.
+	chkDataEnd := chkDataLenStart + n + int(chkDataLen)
+	if chkDataEnd > sgmBytes.Len() {
+		return nil, errors.Errorf("head chunk file doesn't include enough bytes to read the chunk - required:%v, available:%v", chkDataEnd, sgmBytes.Len())
+	}
+
+	// Check the CRC.
+	sum := sgmBytes.Range(chkDataEnd, chkDataEnd+CRCSize)
+	if _, err := chkCRC32.Write(sgmBytes.Range(chkStart-(SeriesRefSize+2*MintMaxtSize), chkDataEnd)); err != nil {
+		return nil, err
+	}
+	if act := chkCRC32.Sum(nil); !bytes.Equal(act, sum) {
+		return nil, errors.Errorf("checksum mismatch expected:%x, actual:%x", sum, act)
+	}
+
+	// The chunk data itself.
+	chkData := sgmBytes.Range(chkDataEnd-int(chkDataLen), chkDataEnd)
+	return cdm.pool.Get(chunkenc.Encoding(chkEnc), chkData)
+}
+
+func (cdm *ChunkDiskMapper) getChunkFromBuffer(chunkRef uint64) chunkenc.Chunk {
+	shardIdx := chunkRef % inBufferShards
+
+	cdm.inBufferChunksLocks[shardIdx].RLock()
+	defer cdm.inBufferChunksLocks[shardIdx].RUnlock()
+
+	return cdm.inBufferChunks[shardIdx][chunkRef]
+}
+
+// IterateAllChunks iterates on all the chunks in its byte slices in the order of the head chunk file sequence
+// and runs the provided function on each chunk. It returns on the first error encountered.
+func (cdm *ChunkDiskMapper) IterateAllChunks(f func(seriesRef, chunkRef uint64, mint, maxt int64) error) error {
+	chkCRC32 := newCRC32()
+
+	// Iterate files in ascending order.
+	seqs := make([]int, 0, len(cdm.mmappedChunkFiles))
+	for seg := range cdm.mmappedChunkFiles {
+		seqs = append(seqs, seg)
+	}
+	sort.Ints(seqs)
+	for _, seq := range seqs {
+		bs := cdm.mmappedChunkFiles[seq]
+		sliceLen := bs.Len()
+		idx := HeadChunkFileHeaderSize
+		for idx < sliceLen {
+			if sliceLen-idx < 30 {
+				return &CorruptionErr{
+					Dir:       cdm.dir.Name(),
+					FileIndex: seq,
+					Err:       errors.Errorf("head chunk file doesn't include enough bytes to read the chunk header - required:%v, available:%v, file:%d", idx+35, sliceLen, seq),
+				}
+			}
+			chkCRC32.Reset()
+			startIdx := idx
+			seriesRef := binary.BigEndian.Uint64(bs.Range(idx, idx+8))
+			idx += 8
+
+			mint := int64(binary.BigEndian.Uint64(bs.Range(idx, idx+8)))
+			idx += 8
+
+			maxt := int64(binary.BigEndian.Uint64(bs.Range(idx, idx+8)))
+			idx += 8
+
+			chunkRef := uint64(seq)<<32 | uint64(idx)
+			if err := f(seriesRef, chunkRef, mint, maxt); err != nil {
```

Comment:
Wait, why we cannot use `Chunk(..)` method for all of this?

Reply:
1. We don't have the ref yet to call `Chunk(...)`
2. We don't need to chunk data here actually, whereas `Chunk(...)` returns the actual chunks. What we need here are the ones passed as an argument, which is not given by `Chunk(...)`.